src/libraries/System.Private.CoreLib/src/System/IO/StringReader.cs

```diff
@@ -175,38 +172,37 @@ public override string ReadToEnd()
         //
         public override string? ReadLine()
         {
-            if (_s == null)
+            string? s = _s;
+            if (s == null)
             {
-                throw new ObjectDisposedException(null, SR.ObjectDisposed_ReaderClosed);
+                ThrowHelper.ThrowObjectDisposedException_ReaderClosed();
             }
+            if ((uint)_pos >= (uint)s.Length)
+                return null;
 
-            int i = _pos;
-            while (i < _length)
+            ReadOnlySpan<char> remaining = s.AsSpan(_pos);
+            int foundLineLength = remaining.IndexOfAny('\r', '\n');
+            if (foundLineLength >= 0)
             {
-                char ch = _s[i];
-                if (ch == '\r' || ch == '\n')
+                string result = foundLineLength == 0
+                    ? string.Empty
+                    : new string(remaining.Slice(0, foundLineLength));
+
+                char ch = remaining[foundLineLength];
+                _pos += foundLineLength + 1;
+                if (ch == '\r' && _pos < s.Length && s[_pos] == '\n')
                 {
-                    string result = _s.Substring(_pos, i - _pos);
-                    _pos = i + 1;
-                    if (ch == '\r' && _pos < _length && _s[_pos] == '\n')
-                    {
-                        _pos++;
-                    }
-
-                    return result;
+                    _pos++;
                 }
 
-                i++;
+                return result;
             }
-
-            if (i > _pos)
+            else
             {
-                string result = _s.Substring(_pos, i - _pos);
-                _pos = i;
+                string result = new string(remaining);
```

Comment:
Today if there's just one line in the input, the `_s.Substring(_pos, i - pos)` will end up special-casing those inputs to just return the original string.  This change loses that.

Reply:
good catch 👍

Reply:
worth adding a test ?

Reply:
if none exists I would agree yes I'll look into it.

Reply:
You'll want `Assert.AreSame` for this test. But otherwise yeah, these two _else_ statements can be collapsed into a single clause for better efficiency. Thanks for looking into it!

Reply:
again want to stay using Spans due to Substring checks, changed single test to use Same

Reply:
Same comment as above - we should make the calling implementation (this code) as simple as possible. Any perceived deficiencies in `Substring` should be addressed directly within `Substring`.